### ISSUE TITLE
AzureFileCopy v4 fails on extracting AzCopy

### ISSUE THREAD

**Issue by gaspo100** [microsoft/azure-pipelines-tasks#14461]
## Note  
Issues in this repo are for tracking bugs, feature requests and questions for the tasks in this repo 

For a list:  
https://github.com/Microsoft/azure-pipelines-tasks/tree/master/Tasks

If you have an issue or request for the Azure Pipelines service, use developer community instead:  

https://developercommunity.visualstudio.com/spaces/21/index.html )

## Required Information

Entering this information will route you directly to the right team and expedite traction.

**Question, Bug, or Feature?**  
*Type*: Bug

**Enter Task Name**: AzureFileCopy v4  

list here (V# not needed):  
https://github.com/Microsoft/azure-pipelines-tasks/tree/master/Tasks

## Environment
- Azure Pipelines
- Agent - Hosted 
    
## Issue Description

When AzureFileCopy v4 downloads AzCopy onto the target VM to folder "C:\ADO_AzCopyV10\AzCopy"  it will fail to extract if the folder already contains content extracted in the previous run.

The issue seems to be with the line 74 in in AzureFileCopyRemoteJob.ps1:
                ```Expand-Archive $azCopyZipPath -DestinationPath $azCopyFolderPath```

It should use -Force argument or clear the folder before extracting.

### Task logs
```
+ CategoryInfo : OperationStopped: (Failed while do...ath in env:Path:String) [], RuntimeException
+ FullyQualifiedErrorId : Failed while downloading azcopy.exe from the URL with exception Failed to create file 
  'C:\ADO_AzCopyV10\AzCopy\AzCopy.exe' while expanding the archive file 'C:\ADO_AzCopyV10\AzCopy.zip' contents as the   
  file 'C:\ADO_AzCopyV10\AzCopy\AzCopy.exe' already exists. Use the -Force parameter if you want to overwrite the ex   
  isting directory 'C:\ADO_AzCopyV10\AzCopy\AzCopy.exe' contents when expanding the archive file.. Please download a    
  zcopy.exe 10.3.3 and set this extracted path in env:Path
```

**Comment by GreatBarrier86** [comment]
We worked around this temporarily by reverting back to V3, but we'd prefer to remain on the latest if possible, especially since V3 doesn't support wildcards.

**Comment by GreatBarrier86** [comment]
@nadesu, is this going to be triaged? It's been waiting for 9 days now. The latest version of Azure File Copy simply doesn't work and that is the default version for all new task entries. This needs to be fixed.

**Comment by gaspo100** [comment]
I temporarily worked around this by deleting the "C:\ADO_AzCopyV10" folder with Power Shell script that is run after the AzCopy task. I also hope that this will be dealt with soon as the fix seems to be as simple as adding -Force to the Expand-Archive command.

**Comment by GreatBarrier86** [comment]
> ... after the AzCopy task. ..

Don't you mean BEFORE the AzCopy task? Or do i misunderstand what you did?

**Comment by gaspo100** [comment]
No, its after. Once the AzCopy task finishes its job copying files onto the tratget VM and I don't need it anymore. 
The next task runs powershell script that deletes the AzCopy folder so the next time the pipeline is run it won't fail extracting the AzCopy.

**Comment by AmrutaKawade** [comment]
Can you share your Task version? is it 4.183.0?

**Comment by gaspo100** [comment]
The task is "Azure file copy" v4.*
![image](https://user-images.githubusercontent.com/931540/110311811-47278780-803f-11eb-9eca-1f9e6b4d0fe3.png)

This task copies AzCopy v10 to the target VM regardless of whether it is already installed there or not.

**Comment by AmrutaKawade** [comment]
> The task is "Azure file copy" v4.*
> ![image](https://user-images.githubusercontent.com/931540/110311811-47278780-803f-11eb-9eca-1f9e6b4d0fe3.png)
> 
> This task copies AzCopy v10 to the target VM regardless of whether it is already installed there or not.

there is no change in logic of task except upgrading version of azcopy exe. 
can you try deleting this folder on VM C:\ADO_AzCopyV10 ? No need to write script to delete it everytime. Deleting just once should work

**Comment by GreatBarrier86** [comment]
@AmrutaKawade, i believe the problem is that the Azure hosted Server 2019 pipeline VM used had ITS version upgraded to 10.8, so the folder is already present when the extraction occurs.

The logic didn’t change but the VM did. That’s the problem.

**Comment by gaspo100** [comment]
In my case the Azure file copy task is part of the release pipeline that copies build pipeline artifacts to the target VM running Windows server 2012, which does not have AzCopy.

Deleting C:\ADO_AzCopyV10 folder once won't help. 
After deploying to the target VM the folder and its content remain on the VM and when the release pipeline is run again it fails to extract AzCopy because it is already there.

As I mention in the bug report, the problem is in the AzureFileCopyRemoteJob.ps1, line 74:
`Expand-Archive $azCopyZipPath -DestinationPath $azCopyFolderPath`
The above line should contine -Force parameter to empty folder before expanding.

Also, there is another problem I reported in another bug #14462.
Azure file copy task adds "C:\ADO_AzCopyV10" to PATH environment variable over and over every time it is run. I need to manually truncate PATH because because it grows too long after a while.

**Comment by AmrutaKawade** [comment]
> Deleting C:\ADO_AzCopyV10 folder once won't help.
> After deploying to the target VM the folder and its content remain on the VM and when the release pipeline is run again it fails to extract AzCopy because it is already there.

I am not able to repro this issue with fresh VM. and it succeeded with every run.
I think this problem occurs when target VM already had older version of azcopy 10.3.3 installed
and based on this logic

https://github.com/microsoft/azure-pipelines-tasks/blob/3da3b2409ea7bb3f950f99b009ec9df033486c42/Tasks/AzureFileCopyV4/AzureFileCopyRemoteJob.ps1#L48

it will set shoulddownload to true and then try to download 10.8 version but while extracting archive its failing since there is already folder available.

if you see here downloading and setting environment variable will happen only when existing version is less than 10.8.0

https://github.com/microsoft/azure-pipelines-tasks/blob/3da3b2409ea7bb3f950f99b009ec9df033486c42/Tasks/AzureFileCopyV4/AzureFileCopyRemoteJob.ps1#L58

so if you delete `C:\ADO_AzCopyV10` folder on target VM and try running release pipeline it will be created with azcopy 10.8.0 version and from next runs it will not fail again. Also it should solve problem of adding environment variable again and again

**Comment by AmrutaKawade** [comment]
> @AmrutaKawade, i believe the problem is that the Azure hosted Server 2019 pipeline VM used had ITS version upgraded to 10.8, so the folder is already present when the extraction occurs. The logic didn’t change but the VM did. That’s the problem. I believe the script needs to take into account that the VM used already has the version installed that it's trying to download.
> 
> ![image](https://user-images.githubusercontent.com/29482393/110318202-35ac9400-7fdb-11eb-9e32-9cefb95efd91.png)

Yes this is what is happening. can you try above workaround

**Comment by GreatBarrier86** [comment]
I tried deleting that directory, but it doesn't exist on the generated VM. I did this in a Powershell script PRIOR to the Copy Files step. Did i not do this correctly?

![image](https://user-images.githubusercontent.com/29482393/110325212-0dc22e00-7fe5-11eb-8030-28fca220fbfc.png)


2021-03-08T12:57:06.4927065Z ##[section]Starting: PowerShell Script
2021-03-08T12:57:06.5059798Z ==============================================================================
2021-03-08T12:57:06.5060117Z Task         : PowerShell
2021-03-08T12:57:06.5060399Z Description  : Run a PowerShell script on Linux, macOS, or Windows
2021-03-08T12:57:06.5060667Z Version      : 2.180.1
2021-03-08T12:57:06.5060918Z Author       : Microsoft Corporation
2021-03-08T12:57:06.5061227Z Help         : https://docs.microsoft.com/azure/devops/pipelines/tasks/utility/powershell
2021-03-08T12:57:06.5061596Z ==============================================================================
2021-03-08T12:57:07.5711630Z Generating script.
2021-03-08T12:57:07.6827214Z ========================== Starting Command Output ===========================
2021-03-08T12:57:07.7078045Z ##[command]"C:\windows\System32\WindowsPowerShell\v1.0\powershell.exe" -NoLogo -NoProfile -NonInteractive -ExecutionPolicy Unrestricted -Command ". 'D:\a\_temp\f70d67f1-176d-42ec-83fb-5350f2c0d385.ps1'"
2021-03-08T12:57:08.1038151Z Remove-Item : Cannot find path 'C:\ADO_AzCopyV10' because it does not exist.
2021-03-08T12:57:08.1038781Z At D:\a\_temp\f70d67f1-176d-42ec-83fb-5350f2c0d385.ps1:3 char:1
2021-03-08T12:57:08.1039216Z + Remove-Item C:\ADO_AzCopyV10 -Force
2021-03-08T12:57:08.1039616Z + ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
2021-03-08T12:57:08.1040335Z     + CategoryInfo          : ObjectNotFound: (C:\ADO_AzCopyV10:String) [Remove-Item], ItemNotFoundException
2021-03-08T12:57:08.1040896Z     + FullyQualifiedErrorId : PathNotFound,Microsoft.PowerShell.Commands.RemoveItemCommand
2021-03-08T12:57:08.1041685Z  
2021-03-08T12:57:08.2162643Z ##[error]PowerShell exited with code '1'.
2021-03-08T12:57:08.2533664Z ##[section]Finishing: PowerShell Script

**Comment by gaspo100** [comment]
I've checked the older logs from release pipelines when it was failing and the task version was
![image](https://user-images.githubusercontent.com/931540/110329538-b4471700-8057-11eb-8825-bd2017575320.png)

When I run the release pipeline now the logs shows version 4.183.0 and it does not fail anymore! I did not know there was an update.

**Comment by AmrutaKawade** [comment]
@gaspo100 4.183.0 is the version when we introduced 10.8 version of azcopy. Great to hear that its working for you.
@GreatBarrier86  this folder is not available on hosted agent but on target VM where you are trying to copy files

![image](https://user-images.githubusercontent.com/65217380/110334664-046aad00-8049-11eb-8cc5-b6364c574633.png)

**Comment by gaspo100** [comment]
Closing, problem fixed with version 4.183.0

**Comment by GreatBarrier86** [comment]
I dont understand. What are we saying is the fix? I'm still getting the error with 4.183.0.

**Comment by AmrutaKawade** [comment]
@GreatBarrier86 what are you selecting as destination type in AzureFileCopy Task ?
Azure Blob or Azure VMs


![image](https://user-images.githubusercontent.com/65217380/110354603-579b2a80-805e-11eb-8976-0c2d9ee13f2d.png)

If you are selecting Azure CM then login to that VM and delete `C:\ADO_AzCopyV10` this folder.

**Comment by GreatBarrier86** [comment]
My mistake. I wasn’t connecting the dots. Is this a one time thing I’ll need to do or every time?

**Comment by AmrutaKawade** [comment]
> My mistake. I wasn’t connecting the dots. Is this a one time thing I’ll need to do or every time?
> 
> this whole time, I was thinking the issue lay with the Azure Pipelines hosted agent VM, not the destination VM. My bad.

Onetime thing

**Comment by antonio-pscanner** [comment]
We just ran into a problem because the azcopy was installed in a different folder than what the task expects on the remote server and was outdated (10.6.0). Solved by executing the same powershell commands in the source code:

`https://github.com/microsoft/azure-pipelines-tasks/blob/36195470336e21a885561c771a8db81371fc5363/Tasks/AzureFileCopyV4/AzureFileCopyRemoteJob.ps1` 

but changing the following lines.
`
  $azCopyFolderName = "ADO_AzCopyV10"
  $azCopyFolderPath = Join-Path -Path $env:systemdrive -ChildPath $azCopyFolderName
`

Is it possible to add a parameter on the task to set this folder?
